### DISCUSSION TITLE
How to report on errors in public collections?

### DISCUSSION THREAD

**Discussion by y-vasyunin** [tkainrad/keycombiner#48] [Q&A]
If I find an error in key shortcuts in some public collection, is there a quick way to correct it?
I’m talking about correcting the original public collection that I don’t have rights to edit.

**Comment by tkainrad** [comment]
Thanks for kicking off the Discussions section!

That is a request that I get quite frequently, happy to get the chance to discuss it publicly. 

I would be very much interested in having KeyCombiner users take part in keeping the public collections up-to-date.
However, there are a few issues that I have to think through first. 

Very often, user-reported errors in public collections turn out to be related to their keyboard layout. There are other sources with lists of keyboard shortcuts on the web, some of them crowd-sourced, but I found that these contain lots of errors.  
With KeyCombiner's public collections, I take great care, often installing the respective software on multiple OSs to test out shortcuts or to look them up if they are only available from an in-app reference.

I do have some hopes to eventually make the public collection data a collaborative open-source project via CSV files in a GitHub repo. I fear though that I will still have to check every user's submission for correctness. 

On the bright side, KeyCombiner users are, as you know, not bound to the data in public collections, they can easily override errors in their personal collection. And, what you can already do is report any errors via mail, the integrated <kbd>Send Feedback</kbd>-button, or through GitHub. Fortunately, this happens quite frequently and I try to be quick with updating the collections.

**Reply by y-vasyunin** [reply]
It’s great to discuss such a helpful app directly with its developer! I think it’s essential to engage your users in collection creation to save your own time.

Omg, I even haven’t thought about different layouts. I use US International on Windows and ABC Extended on Mac. Probably this could have caused several issues I had.

Wouldn’t it be advantageous to have another attribute attached to shortcuts, which would refer to “non-standard” layouts, such as on the screenshot below?

<img width="400" alt="KeyCombiner screenshot" src="https://user-images.githubusercontent.com/18553297/137441905-17565e76-e92c-4f98-abdc-789f5f391a51.png">

**Reply by tkainrad** [reply]
People sometimes wonder if I am not bothered by support requests, given that I build a consumer app as a side-project with very low pricing and even a free tier. On the contrary, I still look forward to every request I receive :)

I do think quite a lot about different layouts. I have written a fairly popular blog post on the mess that is shortcut handling on the web:
https://tkainrad.dev/posts/why-keyboard-shortcuts-dont-work-on-non-us-keyboard-layouts-and-how-to-fix-it/

The post hints at some problems I have with KeyCombiner. Even if I find very sophisticated ways to deal with all possible layouts, this doesn't help if the target application doesn't behave in the same way. If you are supposed to enter a search field via <kbd>/</kbd>, which is <kbd>Shift</kbd>+<kbd>7</kbd> on a German layout, I could make this work in KeyCombiner, but the target application will likely require you to press <kbd>#</kbd> with the said layout.

The good thing with KeyCombiner is that users can just change the combinations to whatever works for them. I.e. on my German keyboard layout, I can just change the above mentions search shortcut from <kbd>/</kbd> to <kbd>#</kbd> and it will behave correctly during practice. It is then even an actual reference for when I need to look up the shortcut. If it still said <kbd>/</kbd>, I would have to think about what this means on my layout.

Still, I am working on a few things to improve support for different layouts. There is already some fairly experimental stuff available in the _Practice Settings_.

I am not sure what you mean by attaching such an attribute to every key combination. What should KeyCombiner do with that information?

**Reply by y-vasyunin** [reply]
Awesome blog post!

> I am not sure what you mean by attaching such an attribute to every key combination. What should KeyCombiner do with that information?

I was referring to some kind of a note, which could be helpful to a user, if he/she uses multiple keyboard layouts. But, honestly, I'm not sure that it's something very necessary. Just disregard it.

**Reply by tkainrad** [reply]
Thanks for the clarification. I will keep the idea in mind, but it will not be on the roadmap for the foreseeable future.
